lib/ntlmauth/ntlmauth.cc

```diff
@@ -108,10 +108,19 @@ ntlm_fetch_string(const ntlmhdr *packet, const int32_t packet_size, const strhdr
     int32_t o = le32toh(str->offset);
```

Comment:
Bad cast here and on the line above it. We should use `auto` to avoid blindly casting unsigned integer to a signed integer. And `const` if possible. Fixing this will require changing code that uses `l` and `o`, of course.

There may be a similar problem with the argument of the function -- the function parameter is unsigned but strhdr::offset is signed. I did _not_ check whether we can change `strhdr` declaration to use unsigned types and whether we at least check `strhdr` field values for not being negative before using them here.

Reply:
The casting effects are intentional and happen prior to input validation so that insane results can be handled per the logic this PR touches.

Reply:
> The casting effects are intentional

AFAICT, until C++20, these effects are implementation defined. In other words, this code relies on an undefined behavior. For example, the compiler has the right to ignore the assignment completely, put a random number into `o`, or call abort(). Here is the corresponding quote from [cppreference](https://en.cppreference.com/w/cpp/language/implicit_conversion):

> If the destination type is signed, the value does not change if the source integer can be represented in the destination type. Otherwise the result is implementation-defined until C++20

Needless to say, this code usually "works" in Squid environment, but it is still a bad cast.